geniza/corpus/templatetags/corpus_extras.py

```diff
@@ -210,6 +217,85 @@ def process_citation(source):
     return mark_safe(Footnote.display_multiple(source.list))
 
 
+def highlight_words(text, highlights):
+    """Given a reference text (the first parameter) and a dictionary of words to be highlighted (the second parameter),
+    this method locates those words in the reference text and surrounds them with css to be highlighted
+    """
+    highlighted_text = text
+    for matched_word in highlights:
+        if not highlights[matched_word] and len(matched_word.strip()) > 0:
+            highlighted_text = highlighted_text.replace(
+                matched_word, f"<em class='search-match'>{matched_word}</em>"
+            )
+            highlights[matched_word] = True
+    return highlighted_text
+
+
+@register.filter
+def find_highlight_regex(reference, query):
+    """Given a reference text (the first parameter) and a search query (the second parameter), this method
+    finds regex matches, highlights them in the reference text, and returns the result.
+    """
+    try:
+        matches = re.findall(query, reference)
+    except:
+        logger.warning(f"Bad regex")
+        return reference
+    return highlight_words(text=reference, highlights=dict.fromkeys(matches, False))
+
+
+@register.filter
+def find_highlight_keywords(reference, query, english=True):
+    """Given a reference text (the first parameter) and a search query (the second parameter), this method
+    finds keywords matches, highlights them in the reference text, and returns the result.
+    """
+
+    import nltk
+
+    from geniza.settings.components.base import NLTK_DATA
+
+    nltk.data.path = [NLTK_DATA]
+    import os.path
+
+    english_stopwords_path = str(
+        nltk.data.path[0] / "corpora" / "stopwords" / "english"
+    )
+    if not os.path.isfile(english_stopwords_path):
+        logger.warning(f"Can't find nltk stopwords in {english_stopwords_path}!")
+        return reference  # TODO mkdir, download, and resume operation
+    from nltk.corpus import stopwords
+    from nltk.stem import PorterStemmer
+    from nltk.tokenize import RegexpTokenizer
+
+    thresh = ENG_MATCH_THRESH if english else ARA_MATCH_THRESH
+    highlights = {}
+    tokenizer = RegexpTokenizer(r"\w+")
+    porter_stemmer = PorterStemmer()
+    query_words = tokenizer.tokenize(query)
+    refined_query_words = [
+        word for word in query_words if word not in stopwords.words("english")
+    ]
+    ref_words = tokenizer.tokenize(reference)
+    refined_ref_words = [
+        word for word in ref_words if word not in stopwords.words("english")
+    ]
+    for word1 in refined_query_words:
+        if len(word1) < 3:
+            continue
+        for word2 in refined_ref_words:
+            if english:
+                stemmed1 = porter_stemmer.stem(word1)
+                stemmed2 = porter_stemmer.stem(word2)
+            else:
+                logger.warning(f"Non-English search not supported yet!")
+                return reference
```

Comment:
Codecov is [complaining](https://app.codecov.io/gh/Princeton-CDH/geniza/pull/1932?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Princeton-CDH) that these two lines aren't covered in unit tests, could you add a test for them? (And I guess this means specifically non-english stemming is not supported, right?)